pkg/controllers/management/auth/globalroles/globalrole_handler_test.go

```diff
@@ -0,0 +1,813 @@
+		Verbs:     []string{"*"},
+		APIGroups: []string{""},
+		Resources: []string{"pod"},
+	}
+
+	namespacedRulesGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespacedRulesGR",
+			UID:  "00000000",
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "fake-kind",
+			APIVersion: "fake-version",
+		},
+		NamespacedRules: map[string][]rbacv1.PolicyRule{
+			"namespace1": {
+				readPodPolicyRule,
+				readConfigPolicyRule,
+			},
+			"namespace2": {
+				adminPodPolicyRule,
+				readPodPolicyRule,
+			},
+		},
+	}
+	namespacedRulesOwnerRef = metav1.OwnerReference{
+		APIVersion: namespacedRulesGR.APIVersion,
+		Kind:       namespacedRulesGR.Kind,
+		Name:       namespacedRulesGR.Name,
+		UID:        namespacedRulesGR.UID,
+	}
+
+	updatedNamespacedRulesGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespacedRulesGR",
+		},
+		NamespacedRules: map[string][]rbacv1.PolicyRule{
+			"namespace1": {
+				readPodPolicyRule,
+			},
+			"namespace2": {
+				adminPodPolicyRule,
+				readPodPolicyRule,
+				readConfigPolicyRule,
+			},
+		},
+	}
+)
+
+type grTestStateChanges struct {
+	t            *testing.T
+	createdRoles map[string]*rbacv1.Role
+	deletedRoles map[string]struct{}
+}
+type grTestState struct {
+	nsListerMock *normanFakes.NamespaceListerMock
+	rListerMock  *rbacFakes.RoleListerMock
+	rClientMock  *rbacFakes.RoleInterfaceMock
+	counter      int
+	stateChanges *grTestStateChanges
+}
+
+func TestReconcileNamespacedRoles(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name            string
+		stateSetup      func(grTestState)
+		stateAssertions func(grTestStateChanges)
+		globalRole      *v3.GlobalRole
+		wantError       bool
+	}{
+		{
+			name: "getting namespace fails",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_, _ string) (*corev1.Namespace, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  true,
+		},
+		{
+			name: "namespace is not found",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_, _ string) (*corev1.Namespace, error) {
+					return nil, apierrors.NewNotFound(schema.GroupResource{
+						Group:    normanv1.RoleGroupVersionKind.Group,
+						Resource: normanv1.RoleGroupVersionResource.Resource,
+					}, "")
+				}
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  false,
+		},
+		{
+			name: "getting role fails",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_, _ string) (*corev1.Namespace, error) {
+					return nil, nil
```

Comment:
I think this (and other such usages) are causing a nil pointer in your tests - your code assumes that if there was no error, there was a namespace, and attempts to retrieve the status, which is nil, producing a panic.